### PR TITLE
Initial documentation of fprime libraries

### PR DESCRIPTION
**Pull request by LeStarch**
| | |
|:---|:---|
|**_Related Issue(s)_**|  |
|**_Has Unit Tests (y/n)_**|  |
|**_Documentation Included (y/n)_**|  |

---
## Change Description

Adds in library how-to guide. 

## Rationale

Needed.  Issue: #2250 

## Future Work

Note any additional work that will be done relating to this issue.

### CODE REVIEW

File: docs/HowTo/develop-fprime-libraries.md [hunk 1]
```diff
@@ -0,0 +1,75 @@
+# How-To: Develop and F´ Library
```
Comment (thomas-bc) [inline docs/HowTo/develop-fprime-libraries.md]: Spelling, `and` -> `an`

File: docs/HowTo/develop-fprime-libraries.md [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+In this section, you will learn about the expected structure of an F´ library. An F´ library must have a `library.cmake` and may have any of the following items:
+
+1. `cmake/toolchain` Folder and Toolchain Files
+2. `cmake/platform` Folder and Platform Files
+3. Module Directories and Modules (Components, Ports, Topologies, etc.)
+
+That means that a complete F´ library might look like the following:
+
+```bash
+library.cmake
```
Comment (thomas-bc) [inline docs/HowTo/develop-fprime-libraries.md]: I would suggest reformatting using some ASCII nicenesses 

```
my-lib/
├── A
│   ├── B
│   └── C
└── library.cmake
```

File: docs/HowTo/develop-fprime-libraries.md [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+add_fprime_subdirectory("${CURRENT_CMAKE_LIST_DIR}/MyLibrary/MyTopology")
+```
+
+`library.cmake` should exist at the root of the library's directory structure.
+
+## Optional: Toolchain Folder and Toolchain Files
+
+F´ libraries often need to supply new toolchain integrations with F´. These libraries must define a `cmake/toolchain` folder and place toolchain files within that folder.  Any file ending in `.cmake` and placed within the `cmake/toolchain` folder will be available as a toolchain via `fprime-util generate <toolchain>`.
+
+The example toolchain above (`cmake/toolchain/ny-toolchain`) would be found as `my-toolchain` in `fprime-util`.
```
Comment (mosa11aei) [inline docs/HowTo/develop-fprime-libraries.md]: Typo: "my-toolchain"

File: docs/HowTo/general.md [hunk 4]
```diff
@@ -6,4 +6,5 @@ This section of the documentation captures how-to guides for various F´ develop
 | Table of Contents                          |
 |--------------------------------------------|
 | [F´ GDS Plugins](./develop-gds-plugins.md) |
+| [F´ Libraries](./develop-fprime-libraries.md) |
```
Comment (thomas-bc) [inline docs/HowTo/general.md:9]: Should also be added to https://github.com/nasa/fprime/blob/devel/docs/_includes/toc.md

File: docs/HowTo/develop-fprime-libraries.md [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+The `cmake/platform` folder may contain any number of platform files and must be placed in the root of the library's directory structure.
+
+## Optional: F´ Module Directories
+
+F´ libraries share F´ code through modules. These module directories are created by `fprime-util new --component` or by hand and may include any F´ code (Components, Ports, Topologies, Data Types, etc.).  These types are made available to users of the library by ensuring that they are added to the `library.cmake` file.  The only restriction is that these component should be placed under a namespacing directory for the library (e.g. `MyLibrary`) to avoid collisions with other libraries and F´ code.
+
+## Conclusion
+
+You should now understand the structure of an F´  library an know how to assemble one.
```
Comment (thomas-bc) [inline docs/HowTo/develop-fprime-libraries.md:85]: I believe this type of `Conclusion` blocks are usually deemed unnecessary/redundant by our designer colleagues

**Review by thomas-bc — CHANGES_REQUESTED** [review]

File: docs/HowTo/develop-fprime-libraries.md [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+4. [Optional: Platform Folder and Platform Files](#optional-platform-folder-and-platform-files)
+5. [Optional: F´ Module Directories](#optional-fmodule-directories)
+
+## F´ Library Structure
+
+In this section, you will learn about the expected structure of an F´ library. An F´ library must have a `library.cmake` and may have any of the following items:
+
+1. `cmake/toolchain` Folder and Toolchain Files
+2. `cmake/platform` Folder and Platform Files
+3. Module Directories and Modules (Components, Ports, Topologies, etc.)
```
Comment (mosa11aei) [inline docs/HowTo/develop-fprime-libraries.md]: Can an example of Ports be included as well? A link can be made to [this file](https://github.com/nasa/fprime/blob/devel/Drv/TcpClient/TcpClient.fpp) as an example of how ports can be segmented across interfaces.

File: docs/HowTo/develop-fprime-libraries.md [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+## Optional: Platform Folder and Platform Files
+
+In a similar manner to toolchains, platforms may be provided in the `cmake/platform` folder. Any file ending in `.cmake` and placed within the `cmake/platform` folder will be available to F´  to be loaded as a platform. Commonly, this means that a toolchain would include the line `set(FPRIME_PLATFORM <platform>)`.
+
+The example platform would thus need to include `set(FPRIME_PLATFORM "my-platform")`.
+
+The `cmake/platform` folder may contain any number of platform files and must be placed in the root of the library's directory structure.
+
+## Optional: F´ Module Directories
```
Comment (mosa11aei) [inline docs/HowTo/develop-fprime-libraries.md]: This section should move higher in this doc, preferably near the structure section.

File: docs/HowTo/develop-fprime-libraries.md [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+
+```bash
+library.cmake
+cmake/toolchain/my-toolchain.cmake
+cmake/platform/my-platform.cmake
+MyLibrary/Components/MyComponent/...
+MyLibrary/MyTopology/...
+```
+
+> Module directories shall be namespaced (i.e. `MyLibrary` above) to ensure that modules do not collide with F´ and other libraries. Placing directories like `Svc` directly at the root of the repository is *strongly* discouraged.
```
Comment (mosa11aei) [inline docs/HowTo/develop-fprime-libraries.md]: I'm confused by what it means to place `Svc` at the root of the repository. As in, moving `Svc` out of `fprime` submodule? Or creating a namespace that is `Svc` outside of F Prime's `Svc`?

**Review by mosa11aei — CHANGES_REQUESTED** [review]
Looks great.

File: docs/HowTo/develop-fprime-libraries.md (hunk 5 above)
Comment (LeStarch) [reply docs/HowTo/develop-fprime-libraries.md:85]: I was told it was encouraged, at lead in HowTo guides.

File: docs/HowTo/develop-fprime-libraries.md (hunk 7 above)
Comment (LeStarch) [reply docs/HowTo/develop-fprime-libraries.md]: Agreed

File: docs/HowTo/develop-fprime-libraries.md (hunk 6 above)
Comment (LeStarch) [reply docs/HowTo/develop-fprime-libraries.md]: I linked to our tutorials which cover how to develop items.

**Comment by LeStarch** [comment]
@alimosallaei19 @thomas-bc I addressed all comments!

File: docs/HowTo/develop-fprime-libraries.md (hunk 5 above)
Comment (thomas-bc) [reply docs/HowTo/develop-fprime-libraries.md:85]: Ok I may have extrapolated from something in the tutorials I think, unsure. It does make sense here in my opinion.

**Review by thomas-bc — APPROVED** [review]

**Pull request merged**